app/src/store/user.ts

```diff
@@ -1,20 +1,107 @@
-import { createSlice } from '@reduxjs/toolkit';
-import type { PayloadAction } from '@reduxjs/toolkit';
+import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
+import config from '../../config';
+import Auth0 from 'react-native-auth0';
+import { type AsyncStoreSlice } from './app';
 
-interface UserState {
+export type UserState = AsyncStoreSlice<{
   accessToken: string;
-}
-
-const initialState = { accessToken: '' } as UserState;
+  email: string;
+}>;
 
 export const userSlice = createSlice({
   name: 'user',
-  initialState,
-  reducers: {
-    setAccessToken: (state, action: PayloadAction<string>) => {
-      state.accessToken = action.payload;
-    },
+  initialState: null as UserState,
+  reducers: {},
+  extraReducers: (builder) => {
+    builder
+      .addCase(checkLoginStatus.pending, () => {
+        return { status: 'pending' };
+      })
+      .addCase(checkLoginStatus.fulfilled, (_, action) => {
+        return { status: 'fulfilled', ...action.payload };
+      })
+      .addCase(checkLoginStatus.rejected, () => {
+        return { status: 'rejected' };
+      });
+    builder
+      .addCase(login.pending, () => {
+        return { status: 'pending' };
+      })
+      .addCase(login.fulfilled, (_, action) => {
+        return { status: 'fulfilled', ...action.payload };
+      })
+      .addCase(login.rejected, (_, action) => {
+        return { status: 'rejected', ...action.error };
+      });
+    builder
+      .addCase(logout.pending, () => {
+        return { status: 'pending' };
+      })
+      .addCase(logout.fulfilled, (_, action) => {
+        return { status: 'fulfilled', ...action.payload };
+      })
+      .addCase(logout.rejected, (_, action) => {
+        return { status: 'rejected', ...action.error };
+      });
   },
 });
 
-export const { setAccessToken } = userSlice.actions;
+const auth0 = new Auth0({
+  domain: config.domain,
+  clientId: config.clientId,
+});
+
+export const checkLoginStatus = createAsyncThunk<{
+  accessToken: string;
+  email: string;
+}>('user/checkLoginStatus', async () => {
+  const credentials = await auth0.credentialsManager.getCredentials();
+  if (credentials) {
+    const user = await auth0.auth.userInfo({
+      token: credentials.accessToken,
+    });
+    return {
+      accessToken: credentials.accessToken,
+      email: user.email,
+    };
+  } else {
+    return { accessToken: '', email: '' };
+  }
+});
+
+export const login = createAsyncThunk<
+  { accessToken: string; email: string },
+  void,
+  { rejectValue: 'ID_TOKEN_UNDEFINED' }
+>('user/login', async (_, { rejectWithValue }) => {
+  const credentials = await auth0.webAuth.authorize({
+    scope: config.auth0Scope,
+    audience: config.audience,
+  });
+
+  if (typeof credentials.idToken === 'undefined') {
+    return rejectWithValue('ID_TOKEN_UNDEFINED');
+  }
+
+  auth0.credentialsManager.saveCredentials(credentials as any);
```

Comment:
<img width="539" alt="Screen Shot 2023-03-13 at 4 17 48 PM" src="https://user-images.githubusercontent.com/35968892/224854244-495019f3-b9bd-457c-a229-8e70d01c6f0b.png">

TS still complains. I think its because this credentials object is coming from auth0 like we discussed before

Reply:
I see, that's strange. Maybe because it's a member of an object, I'm wondering if you just check like this `if (credentials.idToken)` if it would work, but I think it's fine as-is with the `as any` then. LGTM